application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java

```diff
@@ -0,0 +1,303 @@
+            return savedDashboard;
+        } catch (Exception e) {
+            notificationEntityService.notifyEntity(user.getTenantId(), emptyId(EntityType.DASHBOARD), null, null,
+                    actionType, user, e, dashboardId.toString(), customerId.toString());
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public Dashboard assignDashboardToPublicCustomer(DashboardId dashboardId, SecurityUser user) throws ThingsboardException {
+        ActionType actionType = ActionType.ASSIGNED_TO_CUSTOMER;
+        try {
+            Customer publicCustomer = customerService.findOrCreatePublicCustomer(user.getTenantId());
+            Dashboard savedDashboard = checkNotNull(dashboardService.assignDashboardToCustomer(user.getTenantId(), dashboardId, publicCustomer.getId()));
+            notificationEntityService.notifyAssignOrUnassignEntityToCustomer(user.getTenantId(), dashboardId, user.getCustomerId(), savedDashboard,
+                    actionType, null, user, false, dashboardId.toString(),
+                    publicCustomer.getId().toString(), publicCustomer.getName());
+            return savedDashboard;
+        } catch (Exception e) {
+            notificationEntityService.notifyEntity(user.getTenantId(), emptyId(EntityType.DASHBOARD), null, null,
+                    actionType, user, e, dashboardId.toString());
+            throw handleException(e);
+        }
+
+    }
+
+    @Override
+    public Dashboard unassignDashboardFromPublicCustomer(Dashboard dashboard, SecurityUser user) throws ThingsboardException {
+        ActionType actionType = ActionType.UNASSIGNED_FROM_CUSTOMER;
+        try {
+            Customer publicCustomer = customerService.findOrCreatePublicCustomer(dashboard.getTenantId());
+            Dashboard savedDashboard = checkNotNull(dashboardService.unassignDashboardFromCustomer(user.getTenantId(), dashboard.getId(), publicCustomer.getId()));
+            notificationEntityService.notifyAssignOrUnassignEntityToCustomer(user.getTenantId(), dashboard.getId(), user.getCustomerId(), dashboard,
+                    actionType, null, user, false, dashboard.getId().toString(),
+                    publicCustomer.getId().toString(), publicCustomer.getName());
+            return savedDashboard;
+        } catch (Exception e) {
+            notificationEntityService.notifyEntity(user.getTenantId(), emptyId(EntityType.DASHBOARD), null, null,
+                    actionType, user, e, dashboard.getId().toString());
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public Dashboard updateDashboardCustomers(TenantId tenantId, Dashboard dashboard, String[] strCustomerIds, SecurityUser user) throws ThingsboardException {
+        ActionType actionType = ActionType.ASSIGNED_TO_CUSTOMER;
+        try {
+
+
+            Set<CustomerId> customerIds = new HashSet<>();
```

Comment:
you should do this in controller

Reply:
Disagree, the maximum number of operations should be removed from the controller to the service